torchtune/modules/transformer.py

```diff
@@ -90,13 +99,147 @@ def forward(
+        self.attn = attn
+        self.mlp = mlp
+        self.ca_norm = ca_norm or nn.Identity()
+        self.mlp_norm = mlp_norm or nn.Identity()
+        self.ca_scale = ca_scale or nn.Identity()
+        self.mlp_scale = mlp_scale or nn.Identity()
+
+    def setup_cache(self, batch_size: int, dtype: torch.dtype) -> None:
+        """Setup key value caches for attention calculation.
+
+        Args:
+            batch_size (int): batch size for the caches.
+            dtype (torch.dtype): dtype for the caches.
+        """
+        self.attn.setup_cache(batch_size, dtype)
+
+    @property
+    def cache_enabled(self) -> bool:
+        """Check if the key value caches are setup."""
+        return self.attn.kv_cache is not None
+
+    def reset_cache(self):
+        """Reset the key value caches."""
+        self.attn.reset_cache()
+
+    def _skip_mask(self, mask: Optional[Tensor]) -> Optional[Tensor]:
+        """Some tokens in x may not attend to any encoder inputs
+        due to the cross attention mask (encoder_mask). This results in
+        a full row of the attention matrix being masked out.
+
+        In the example below, the word "the" is masked from every embedding.
+
+        .. code-block:: text
+
+            |emb||emb||emb|
+        |The| x    x    x
+        |red|      x
+        |car| x
+
+        This results in no inputs into the softmax layer which causes a NaN.
+        The skip mask removes the output of the attention module and
+        mlp resulting in the token being skipped.
+
+        """
+        if mask is None:
+            return None
+        if mask.dtype == torch.bool:
+            mask = ~mask
+        else:
+            mask = torch.isneginf(mask)
+        mask = torch.all(mask, dim=-1, keepdim=True)
+        return mask
+
+    def forward(
+        self,
+        x: Tensor,
+        *,
+        encoder_input: Optional[Tensor] = None,
+        encoder_mask: Optional[Tensor] = None,
+        **kwargs: Dict,
+    ) -> Tensor:
+        """
+        Args:
+            x (Tensor): input tensor with shape
+                [batch_size x seq_length x embed_dim]
+            encoder_input (Optional[Tensor]): Optional second input to cross attend with x. Shape
+                [batch_size x seq_length x embed_dim] (seq_length and embed_dim may very from x)
+            encoder_mask (Optional[Tensor]):  Cross attention boolean tensor with shape
+                [batch_size x x_seq_len x encoder_seq_len]
+            **kwargs (Dict): transformer layer inputs not relevant to self attention.
+
+        Returns:
+            Tensor: output tensor with same shape as input
+                [batch_size x seq_length x embed_dim]
+        """
+        # During decoding, it's possible encoder_input is None because the embeds
+        # are already stored in the kv cache.
+        empty_cache = not self.cache_enabled or self.attn.kv_cache.size == 0
```

Comment:
I don't think I follow?

Reply:
You're supposed to skip the cross attention layer if there is no encoder input since there's nothing to cross attend to. But during decoding, there might be no encoder_input because it's already cached. So this checks both that the cache is empty and that there's no current input.